tests/host_tools/fcmetrics.py

```diff
@@ -363,3 +367,126 @@ def validate_per_device_metrics(self, fc_metrics):
+            full_key = f"{group}.{key}"
+            # values are 0 when:
+            # - there is no update
+            # - device is not used
+            # - SharedIncMetric reset to 0 on flush so if
+            #   there is no change metric the values remain 0.
+            # We can save the amount of bytes we export to
+            # CloudWatch in these cases.
+            # The other cases of 0 are failures in which case
+            # we should assert instead of sending to CW.
+            if (
+                "fail" in full_key
+                or "err" in full_key
+                or "num_faults" in full_key
+                or "panic" in full_key
+            ) and assert_on_err is True:
+                assert 0 == value, f"Error reported in metrics: {full_key}={value}"
+
+            if value != 0:
+                if full_key in skip:
+                    continue
+                unit = get_emf_unit_for_fc_metrics(key)
+                metrics.put_metric(f"fc_metrics.{full_key}", value, unit=unit)
+                # print(f"fc_metrics.{full_key} {value}, unit={unit}")
+                if full_key in metrics_to_export_once:
+                    skip.add(full_key)
+
+
+class FCMetricsMonitor(Thread):
+    """
+    read Firecracker metrics from the microvm every 60secs and
+    uploads the metrics to CW.
+    We do this as a daemon thread every 60sec, instead of collecting
+    all metrics together in the end, to retain timestamp of the metrics.
+    """
+
+    def __init__(self, vm, dimensions, assert_on_err=False):
+        Thread.__init__(self, daemon=True)
+        self.vm = vm
+        self.dimensions = dimensions
+        self.assert_on_err = assert_on_err
+
+        self.metrics_index = 0
+        self.running = True
+
+        self.metrics_logger = get_metrics_logger()
+        self.metrics_logger.set_dimensions(dimensions)
+
+    def _flush_metrics(self):
+        """
+        Since vm.flush_metrics provides only the latest metrics,
+        we call vm.get_all_metrics() instead to be able to collect
+        and upload all metrics data emitted by the microvm.
+        This utility function is created to keep common code in one
+        place and is called every 60sec once the daemon starts and then
+        once when the daemon stops.
+        """
+        metrics = self.vm.get_all_metrics()
+        for index in range(self.metrics_index, len(metrics)):
+            flush_fc_metrics_to_cw(metrics[index], self.metrics_logger)
+            self.metrics_index += 1
+
+    def stop(self):
+        """
+        Stop the daemon gracefully.
+        Since we depend on the vm to provide the metrics,
+        this method should be called just before killing the vm.
+        We collect final metrics here in stop instead of letting it
+        be collected from the "run" method because, "run" could be
+        in sleep when stop is called and once it wakes out of sleep
+        the "vm" might not be avaiable to provide the metrics.
```

Comment:
I think there could be a race here if both threads are running, since stop will be called from a different thread. Maybe Python's GIL will avoid that case, but I think there's still a chance we submit the last metric twice.

Could we just kill the thread?

Reply:
I thought so too but couldn't find a better way to do it, I'll try the kill approach locally to see how it works.

Reply:
instead of killing the thread I tried to work around it by making the stop wait for 1 sec before collecting metrics and shortened the sleep of run() to 1 sec so that it can wake up sooner and break the loop

Reply:
Makes sense, thanks!